File: cpp/tests/test_autotune_filter_gate_evaluator.cpp
// SPDX-License-Identifier: MIT
//
// doctest unit tests for tuner_core::autotune_filter_gate_evaluator.

#include "doctest.h"

#include "tuner_core/autotune_filter_gate_evaluator.hpp"

#include <vector>

using namespace tuner_core::autotune_filter_gate_evaluator;
using ValueMap = tuner_core::sample_gate_helpers::ValueMap;

namespace {

Gate make_gate(std::string name) {
    Gate g;
    g.name = std::move(name);
    g.default_enabled = true;
    return g;
}

Gate make_parametric(std::string name, std::string channel, std::string op, double thr) {
    Gate g;
    g.name = std::move(name);
    g.channel = std::move(channel);
    g.op = std::move(op);
    g.threshold = thr;
    g.default_enabled = true;
    return g;
}

}  // namespace

TEST_CASE("disabled-by-default gate is a pass-through") {
    Gate g = make_parametric("custom", "rpm", ">", 5000.0);
    g.default_enabled = false;
    ValueMap v{{"rpm", 6000.0}};  // would otherwise reject
    auto r = evaluate(g, v);
    CHECK(r.accepted == true);
}

TEST_CASE("std_Custom is a pass-through even when enabled") {
    auto g = make_gate("std_Custom");
    ValueMap v;
    auto r = evaluate(g, v);
    CHECK(r.accepted == true);
}

TEST_CASE("std_DeadLambda accepts a plausible lambda reading") {
    auto g = make_gate("std_DeadLambda");
    ValueMap v{{"lambda1", 1.05}};
    auto r = evaluate(g, v);
    CHECK(r.accepted == true);
}

TEST_CASE("std_DeadLambda rejects when lambda is missing") {
    auto g = make_gate("std_DeadLambda");
    ValueMap v{{"rpm", 5500.0}};
    auto r = evaluate(g, v);
    CHECK(r.accepted == false);
    CHECK(r.reason == "no lambda/AFR channel present");
}

TEST_CASE("std_DeadLambda rejects when lambda is out of range") {
    auto g = make_gate("std_DeadLambda");
    ValueMap v{{"lambda", 0.3}};
    auto r = evaluate(g, v);
    CHECK(r.accepted == false);
    CHECK(r.reason.find("outside plausible range") != std::string::npos);
}

TEST_CASE("std_DeadLambda derives lambda from AFR") {
    auto g = make_gate("std_DeadLambda");
    ValueMap v{{"afr1", 14.7}};  // → lambda 1.0
    auto r = evaluate(g, v);
    CHECK(r.accepted == true);
}

TEST_CASE("std_xAxisMin rejects below the minimum") {
    auto g = make_gate("std_xAxisMin");
    AxisContext axis;
    axis.x_value = 100.0;
    axis.x_min = 200.0;
    ValueMap v;
    auto r = evaluate(g, v, &axis);
    CHECK(r.accepted == false);
    CHECK(r.reason.find("X value 100.0 below axis min 200.0") != std::string::npos);
}

TEST_CASE("std_xAxisMin passes when no axis context is supplied") {
    auto g = make_gate("std_xAxisMin");
    ValueMap v;
    auto r = evaluate(g, v, nullptr);
    CHECK(r.accepted == true);
}

TEST_CASE("std_yAxisMax rejects above the maximum") {
    auto g = make_gate("std_yAxisMax");
    AxisContext axis;
    axis.y_value = 250.0;
    axis.y_max = 200.0;
    ValueMap v;
    auto r = evaluate(g, v, &axis);
    CHECK(r.accepted == false);
    CHECK(r.reason.find("above axis max") != std::string::npos);
}

TEST_CASE("parametric gate rejects when reject condition fires") {
    auto g = make_parametric("minRPM", "rpm", "<", 300.0);
    ValueMap v{{"rpm", 200.0}};
    auto r = evaluate(g, v);
    CHECK(r.accepted == false);
    CHECK(r.reason.find("rpm=200.0") != std::string::npos);
}

TEST_CASE("parametric gate passes when channel is missing from record") {
    auto g = make_parametric("minRPM", "rpm", "<", 300.0);
    ValueMap v;
    auto r = evaluate(g, v);
    CHECK(r.accepted == true);
}

TEST_CASE("evaluate_all stops at the first rejection when fail_fast=true") {
    std::vector<Gate> gates{
        make_parametric("minRPM", "rpm", "<", 300.0),
        make_parametric("maxRPM", "rpm", ">", 7000.0),
    };
    ValueMap v{{"rpm", 200.0}};
    auto results = evaluate_all(gates, v, nullptr, true);
    REQUIRE(results.size() == 1);
    CHECK(results[0].accepted == false);
}

TEST_CASE("evaluate_all evaluates all gates when fail_fast=false") {
    std::vector<Gate> gates{
        make_parametric("minRPM", "rpm", "<", 300.0),
        make_parametric("maxRPM", "rpm", ">", 7000.0),
    };
    ValueMap v{{"rpm", 200.0}};
    auto results = evaluate_all(gates, v, nullptr, false);
    REQUIRE(results.size() == 2);
    CHECK(results[0].accepted == false);
    CHECK(results[1].accepted == true);
}

TEST_CASE("gate_label returns explicit label when set") {
    Gate g = make_gate("std_DeadLambda");
    g.label = "My Custom Label";
    CHECK(gate_label(g) == "My Custom Label");
}

TEST_CASE("gate_label returns standard-gate label for known names") {
    Gate g = make_gate("std_xAxisMin");
    CHECK(gate_label(g) == "Below X-axis minimum");
}

TEST_CASE("gate_label falls back to gate name") {
    Gate g = make_gate("customGate");
    CHECK(gate_label(g) == "customGate");
}

File: tests/unit/test_autotune_filter_gate_evaluator.py
"""Tests for AutotuneFilterGateEvaluator.

Covers:
- Standard named gates (std_DeadLambda, std_xAxisMin/Max, std_yAxisMin/Max, std_Custom)
- Parametric gates with all supported operators (<, >, <=, >=, ==, !=, &)
- Default-disabled gates pass through without evaluating
- Missing channel pass-through
- Operator normalisation (= treated as ==)
"""

from __future__ import annotations

import pytest

from tuner.domain.ecu_definition import AutotuneFilterGate
from tuner.services.autotune_filter_gate_evaluator import (
    AutotuneFilterGateEvaluator,
    AxisContext,
)


def _gate(
    name: str,
    label: str | None = None,
    channel: str | None = None,
    operator: str | None = None,
    threshold: float | None = None,
    default_enabled: bool = True,
) -> AutotuneFilterGate:
    return AutotuneFilterGate(
        name=name,
        label=label,
        channel=channel,
        operator=operator,
        threshold=threshold,
        default_enabled=default_enabled,
    )


# ---------------------------------------------------------------------------
# std_DeadLambda
# ---------------------------------------------------------------------------

class TestStdDeadLambda:
    _GATE = _gate("std_DeadLambda")

    def test_rejects_high_afr(self) -> None:
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(self._GATE, {"afr": 40.0})
        assert not result.accepted
        assert "lambda" in result.reason.lower() or "40" in result.reason

    def test_rejects_missing_channel(self) -> None:
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(self._GATE, {"rpm": 1000.0})
        assert not result.accepted
        assert "no lambda" in result.reason

    def test_accepts_normal_afr(self) -> None:
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(self._GATE, {"afr": 14.7})
        assert result.accepted

    def test_accepts_lambda_channel(self) -> None:
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(self._GATE, {"lambda": 1.0})
        assert result.accepted

    def test_rejects_lambda_out_of_range(self) -> None:
        ev = AutotuneFilterGateEvaluator()
        # lambda 0.3 = AFR ~4.4 — implausible
        result = ev.evaluate(self._GATE, {"lambda": 0.3})
        assert not result.accepted

    def test_accepts_boundary_lambda(self) -> None:
        ev = AutotuneFilterGateEvaluator()
        assert AutotuneFilterGateEvaluator().evaluate(self._GATE, {"lambda": 0.5}).accepted
        assert AutotuneFilterGateEvaluator().evaluate(self._GATE, {"lambda": 1.8}).accepted


# ---------------------------------------------------------------------------
# std_Custom pass-through
# ---------------------------------------------------------------------------

def test_std_custom_always_passes() -> None:
    gate = _gate("std_Custom")
    ev = AutotuneFilterGateEvaluator()
    result = ev.evaluate(gate, {"afr": 40.0})
    assert result.accepted


# ---------------------------------------------------------------------------
# Axis bound gates
# ---------------------------------------------------------------------------

class TestAxisBoundGates:
    def test_x_axis_min_rejects(self) -> None:
        gate = _gate("std_xAxisMin")
        ev = AutotuneFilterGateEvaluator()
        ctx = AxisContext(x_value=400.0, x_min=500.0)
        result = ev.evaluate(gate, {}, axis_context=ctx)
        assert not result.accepted
        assert "400" in result.reason

    def test_x_axis_min_accepts(self) -> None:
        gate = _gate("std_xAxisMin")
        ev = AutotuneFilterGateEvaluator()
        ctx = AxisContext(x_value=600.0, x_min=500.0)
        result = ev.evaluate(gate, {}, axis_context=ctx)
        assert result.accepted

    def test_x_axis_max_rejects(self) -> None:
        gate = _gate("std_xAxisMax")
        ev = AutotuneFilterGateEvaluator()
        ctx = AxisContext(x_value=7000.0, x_max=6000.0)
        result = ev.evaluate(gate, {}, axis_context=ctx)
        assert not result.accepted

    def test_y_axis_min_rejects(self) -> None:
        gate = _gate("std_yAxisMin")
        ev = AutotuneFilterGateEvaluator()
        ctx = AxisContext(y_value=20.0, y_min=30.0)
        result = ev.evaluate(gate, {}, axis_context=ctx)
        assert not result.accepted

    def test_y_axis_max_rejects(self) -> None:
        gate = _gate("std_yAxisMax")
        ev = AutotuneFilterGateEvaluator()
        ctx = AxisContext(y_value=110.0, y_max=100.0)
        result = ev.evaluate(gate, {}, axis_context=ctx)
        assert not result.accepted

    def test_axis_gate_passes_without_context(self) -> None:
        for gate_name in ("std_xAxisMin", "std_xAxisMax", "std_yAxisMin", "std_yAxisMax"):
            gate = _gate(gate_name)
            ev = AutotuneFilterGateEvaluator()
            result = ev.evaluate(gate, {"rpm": 1000.0})
            assert result.accepted, f"{gate_name} should pass through without axis context"


# ---------------------------------------------------------------------------
# Parametric gates — all operators
# ---------------------------------------------------------------------------

class TestParametricGates:
    def test_less_than_rejects(self) -> None:
        gate = _gate("minCltFilter", "Minimum CLT", "coolant", "<", 71.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"coolant": 50.0})
        assert not result.accepted
        assert "coolant" in result.reason

    def test_less_than_accepts(self) -> None:
        gate = _gate("minCltFilter", "Minimum CLT", "coolant", "<", 71.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"coolant": 85.0})
        assert result.accepted

    def test_less_than_boundary(self) -> None:
        gate = _gate("minCltFilter", "Minimum CLT", "coolant", "<", 71.0)
        ev = AutotuneFilterGateEvaluator()
        assert ev.evaluate(gate, {"coolant": 71.0}).accepted

    def test_greater_than_rejects(self) -> None:
        gate = _gate("maxTPS", "Max TPS", "throttle", ">", 15.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"tps": 20.0})
        assert not result.accepted

    def test_greater_than_accepts(self) -> None:
        gate = _gate("maxTPS", "Max TPS", "throttle", ">", 15.0)
        ev = AutotuneFilterGateEvaluator()
        assert ev.evaluate(gate, {"tps": 10.0}).accepted

    def test_equal_rejects(self) -> None:
        gate = _gate("overrunFilter", "Overrun", "pulsewidth", "=", 0.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"pulseWidth": 0.0})
        assert not result.accepted

    def test_equal_accepts(self) -> None:
        gate = _gate("overrunFilter", "Overrun", "pulsewidth", "=", 0.0)
        ev = AutotuneFilterGateEvaluator()
        assert ev.evaluate(gate, {"pulseWidth": 3.5}).accepted

    def test_double_equal_normalised(self) -> None:
        gate = _gate("overrunFilter", "Overrun", "pulsewidth", "==", 0.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"pulseWidth": 0.0})
        assert not result.accepted

    def test_not_equal_rejects(self) -> None:
        gate = _gate("syncCheck", "Sync check", "sync", "!=", 1.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"sync": 0.0})
        assert not result.accepted

    def test_not_equal_accepts(self) -> None:
        gate = _gate("syncCheck", "Sync check", "sync", "!=", 1.0)
        ev = AutotuneFilterGateEvaluator()
        assert ev.evaluate(gate, {"sync": 1.0}).accepted

    def test_less_than_equal_rejects_at_boundary(self) -> None:
        gate = _gate("test", "Test <=", "rpm", "<=", 300.0)
        ev = AutotuneFilterGateEvaluator()
        assert not ev.evaluate(gate, {"rpm": 300.0}).accepted

    def test_greater_than_equal_rejects_at_boundary(self) -> None:
        gate = _gate("test", "Test >=", "rpm", ">=", 6000.0)
        ev = AutotuneFilterGateEvaluator()
        assert not ev.evaluate(gate, {"rpm": 6000.0}).accepted

    def test_bitwise_and_rejects(self) -> None:
        # accelFilter: engine & 16 → reject if accel enrichment active
        gate = _gate("accelFilter", "Accel Flag", "engine", "&", 16.0)
        ev = AutotuneFilterGateEvaluator()
        result = ev.evaluate(gate, {"engine": 16.0})
        assert not result.accepted
        assert "accelFilter" in result.reason or "Accel" in result.reason

    def test_bitwise_and_accepts_when_bit_clear(self) -> None:
        gate = _gate("accelFilter", "Accel Flag", "engine", "&", 16.0)
        ev = AutotuneFilterGateEvaluator()
        assert ev.evaluate(gate, {"engine": 0x01}).accepted

    def test_bitwise_and_rejects_combined_flags(self) -> None:
        gate = _gate("aseFilter", "ASE Flag", "engine", "&", 4.0)
        ev = AutotuneFilterGateEvaluator()
        assert not ev.evaluate(gate, {"engine": 0x07}).accepted


# ---------------------------------------------------------------------------
# Default-disabled gate passes through
# ---------------------------------------------------------------------------

def test_disabled_gate_always_passes() -> None:
    gate = _gate("accelFilter", "Accel Flag", "engine", "&", 16.0, default_enabled=False)
    ev = AutotuneFilterGateEvaluator()
    result = ev.evaluate(gate, {"engine": 16.0})
    assert result.accepted


# ---------------------------------------------------------------------------
# Missing channel pass-through
# ---------------------------------------------------------------------------

def test_parametric_gate_passes_through_when_channel_missing() -> None:
    gate = _gate("minCltFilter", "Min CLT", "coolant", "<", 71.0)
    ev = AutotuneFilterGateEvaluator()
    result = ev.evaluate(gate, {"rpm": 1000.0})  # no coolant channel
    assert result.accepted


# ---------------------------------------------------------------------------
# evaluate_all
# ---------------------------------------------------------------------------

class TestEvaluateAll:
    def test_fail_fast_stops_at_first_rejection(self) -> None:
        gates = (
            _gate("minCltFilter", "Min CLT", "coolant", "<", 71.0),
            _gate("overrunFilter", "Overrun", "pulsewidth", "=", 0.0),
        )
        ev = AutotuneFilterGateEvaluator()
        results = ev.evaluate_all(
            gates,
            {"coolant": 50.0, "pulseWidth": 0.0},
            fail_fast=True,
        )
        assert len(results) == 1
        assert not results[0].accepted
        assert results[0].gate_name == "minCltFilter"

    def test_no_fail_fast_evaluates_all(self) -> None:
        gates = (
            _gate("minCltFilter", "Min CLT", "coolant", "<", 71.0),
            _gate("overrunFilter", "Overrun", "pulsewidth", "=", 0.0),
        )
        ev = AutotuneFilterGateEvaluator()
        results = ev.evaluate_all(
            gates,
            {"coolant": 50.0, "pulseWidth": 0.0},
            fail_fast=False,
        )
        assert len(results) == 2
        assert all(not r.accepted for r in results)

    def test_all_accepted(self) -> None:
        gates = (
            _gate("minCltFilter", "Min CLT", "coolant", "<", 71.0),
            _gate("overrunFilter", "Overrun", "pulsewidth", "=", 0.0),
        )
        ev = AutotuneFilterGateEvaluator()
        results = ev.evaluate_all(
            gates,
            {"coolant": 85.0, "pulseWidth": 3.5},
        )
        assert len(results) == 2
        assert all(r.accepted for r in results)


# ---------------------------------------------------------------------------
# gate_label
# ---------------------------------------------------------------------------

def test_gate_label_returns_ini_label_when_present() -> None:
    gate = _gate("minCltFilter", "Minimum CLT", "coolant", "<", 71.0)
    ev = AutotuneFilterGateEvaluator()
    assert ev.gate_label(gate) == "Minimum CLT"


def test_gate_label_falls_back_to_standard_label() -> None:
    gate = _gate("std_DeadLambda")
    ev = AutotuneFilterGateEvaluator()
    assert "lambda" in ev.gate_label(gate).lower() or "dead" in ev.gate_label(gate).lower()


def test_gate_label_falls_back_to_name() -> None:
    gate = _gate("customGate123")
    ev = AutotuneFilterGateEvaluator()
    assert ev.gate_label(gate) == "customGate123"
